x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.esql.evaluator.predicate.operator.comparison;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.function.scalar.BinaryScalarFunction;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+
+public abstract class InsensitiveBinaryComparison extends BinaryScalarFunction {
+
+    protected InsensitiveBinaryComparison(Source source, Expression left, Expression right) {
+        super(source, left, right);
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.BOOLEAN;
+    }
+
+    public static boolean isWildcard(BytesRef value) {
```

Comment:
You probably already ruled this one out, but I'm mentioning it nevertheless.
In EQL the Optimizer is doing something that is in other few places where a rule is rewriting a specific operation with something that already exists in a different form already in the language. In SQL there is `Optimizer.ReplaceAggregatesWithLiterals` where `sum(1)` is replaced with an `iif`. For EQL it's `Optimizer.ReplaceWildcards` where the InsensitiveBinaryComparison logic is the same as the `Like` one.

So, the logic in ESQL for `=~` couldn't be, in part at least, replaced with what we already have for `LIKE`? We have a `RegexMatchEvaluator`, we have the `WildcardLike` that behaves exactly like `field_name =~ "*foo*"` so we could reuse this logic, no? With the Optimizer doing the replacement for `=~` in case the right hand side has wildcards, into a WildcardLike construct.

Reply:
TBH I think it should be the other way around, ie. 
- do not allow wildcards in `=~` (or better, do not treat them as wildcards, same as `==`)
- implement `LIKE~` (with wildcard support of course)
- replace `LIKE~` with `=~` if the pattern does not include wildcards, same as we already do for `LIKE` and `==`

Supporting wildcards in `=~` would make `LIKE~` completely redundant (there would be no difference between the two in terms of functionality). 
Moreover, if we support wildcards in `=~`, we will force users to escape wildcards in some cases (`?` in particular, that could be pretty common in text fields that contain natural language), and we know that escaping is painful.

Reply:
Your review is in line with the scope of this PR and with the original issue.
I just wanted to raise this concern here before moving forward.

I'd really like to have an opinion on the practical points (regardless of EQL, that is out of scope), in particular:
- what is the advantage of having two operators (`=~` and `LIKE~`) that do the exact same thing?
- we have `==` for exact match and `LIKE` for wildcard match, wouldn't it be more consistent to have `=~` for insensitive exact match and `LIKE~` for insensitive wildcard match?
- we have a rule that translates `LIKE` to `==` if there are no wildcards, why don't we do the same for `LIKE~` and `=~`?
- do we really want to force end users to escape `?` for exact match? 
  - for `==`: `{"query":"from idx | where x == \"how are you?\""}`
  - for `=~`: `{"query":"from idx | where x == \"how are you\\\\?\""}`

Reply:
@luigidellaquila please keep this PR focused on the items that are clear and move the rest of the items into follow-ups.
Namely let's trim this PR on just the first item in #103599:

> Introduce =~ operator for performing case insensitive equality between a string field and a string literal
The field can be both single and multi value. The literal has to be single value.

The multi-value aspect was descoped so let's get this PR to do:
> Introduce =~ operator for performing case insensitive equality between a string field and a string literal.
> The literal has to be single value.

Let's keep the wildcard, like~, etc.. for later.
Most of the code is already in - please clean it up, beef the testing, polish it up and then let's give it a final round of reviews.

For questions, let's talk offline.

Reply:
> Let's keep the wildcard, like~, etc.. for later.

I simplified the code removing wildcard logic. Now the scope is much more limited and IMHO it's self-contained enough, coherent with other operators (`==` logic in particular), and a good starting point for further evolution.

I added a few more tests, although the testable surface is much smaller with this simplification